src/compose/ComposeBox.js

```diff
@@ -217,6 +219,36 @@ class ComposeBox extends PureComponent<Props, State> {
     }
   };
 
+  replaceMessageTextAtRange = (text: string, beg: number, end: number) => {
+    const { message } = this.state;
+    this.setMessageInputValue(message.slice(0, beg) + text + message.slice(end, message.length));
+    return [beg, beg + text.length];
+  };
+
+  insertAttachment = async (attachments: [DocumentPickerResponse]) => {
+    this.setState(({ isUploading }) => ({
+      isUploading: !isUploading,
+    }));
+
+    const { _, auth } = this.props;
+    let [beg, end] = [undefined, this.state.selection.start];
+    for (let i = 0; i < attachments.length; i++) {
+      const fileName = attachments[i].name ?? `attachment-${i + 1}`;
+      [beg, end] = this.replaceMessageTextAtRange(`[Uploading ${fileName}...]()\n`, end, end);
+      try {
+        const response = await api.uploadFile(auth, attachments[i].uri, fileName);
+        [beg, end] = this.replaceMessageTextAtRange(`[${fileName}](${response.uri})\n\n`, beg, end);
+      } catch (e) {
+        [beg, end] = this.replaceMessageTextAtRange('', beg, end);
+        showToast(_(`Uploading ${fileName} unsuccessful`));
+      }
+    }
+
+    this.setState(({ isUploading }) => ({
+      isUploading: !isUploading,
+    }));
+  };
```

Comment:
But then the counterpart to just setting to `true` above would be to just set to `false` here. And that isn't quite right either -- if the first upload is taking a while, the user could potentially start a second one at the same time, and the second could even finish before the first one.

(The `!isUploading` version doesn't correctly handle that case either: it'll set it to false when the second upload starts.)

I think a version that works correctly would be: instead of a boolean, have a `num` as a counter. At the start of the function, do:
```js
this.setState(({ numUploading }) => ({ numUploading: numUploading + 1 }));
```

At the end of the function, do the reverse:
```js
this.setState(({ numUploading }) => ({ numUploading: numUploading - 1 }));
```

And then at render time when we want to know if we're currently uploading, check whether `numUploading > 0`.

Reply:
did the necessary changes.